src/Mvc/src/Microsoft.AspNetCore.Mvc.ViewFeatures/Infrastructure/HttpUriHelper.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Components.Services;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Extensions;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures
+{
+    internal class HttpUriHelper : UriHelperBase
+    {
+        private readonly HttpContext _context;
+
+        public HttpUriHelper(HttpContext context)
+        {
+            _context = context;
+        }
+
+        protected override void InitializeState()
+        {
+            SetAbsoluteBaseUri(GetContextBaseUri());
+            SetAbsoluteUri(GetFullUri());
+        }
+
+        private string GetFullUri()
+        {
+            var request = _context.Request;
+            return UriHelper.BuildAbsolute(
+                request.Scheme,
+                request.Host,
+                request.PathBase,
+                request.Path,
+                request.QueryString);
+        }
+
+        private string GetContextBaseUri()
+        {
+            var request = _context.Request;
+            return UriHelper.BuildAbsolute(request.Scheme, request.Host, request.PathBase);
+        }
+
+        protected override void NavigateToCore(string uri, bool forceLoad)
+        {
+            // For now throw.
+            // We don't have a good way to do redirects on a prerrendering context.
+            // Currently there's no good way to stop the rendering process. We can fix
+            // this by passing in a cancellation token to RenderRootComponentAsync and
+            // having that checked before we await on every ProcessAsynchronousWork iteration
```

Comment:
Cancelling on redirection would be a nice-to-have, but even if we don't do that, it would still be possible to send the redirection response and then ignore the subsequent completion of async prerendering.